pkg/sql/schema_changer_test.go

```diff
@@ -8286,6 +8286,46 @@ func TestVirtualColumnNotAllowedInPkeyBefore22_1(t *testing.T) {
 	require.Equal(t, "pq: cannot use virtual column \"b\" in primary key", err.Error())
 }
 
+func TestHashShardedIndexNotSupportedInMixedVersion(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	params, _ := tests.CreateTestServerParams()
+	params.Knobs.Server = &server.TestingKnobs{
+		DisableAutomaticVersionUpgrade: make(chan struct{}),
+		BinaryVersionOverride:          clusterversion.ByKey(clusterversion.V21_2),
+	}
+
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
+
+	expectedErrMsg := "pq: cannot create hash sharded index on a node running binary newer than 21.2 in a mixed version cluster " +
+		"during a cluster upgrade from pre-22.1. Please run the schema change on a node still running pre-22.1 binary " +
+		"or wait until the cluster upgrade finalized to 22.1. Hash sharded indexes have been improved in 22.1 to use " +
+		"a virtual column and can be added more efficiently"
```

Comment:
nit: consider defining your error message as a `const` in `create_index.go` or use `require.Regexp` instead.